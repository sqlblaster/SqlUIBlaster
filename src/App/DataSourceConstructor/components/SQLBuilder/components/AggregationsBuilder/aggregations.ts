export enum AggregationOperatorEnum {
  'Count of rows',
  'Sum of',
  'Average of',
  'Number of distinct values of',
  'Cumulative sum of',
  'Cumulative count of rows',
  'Standard deviation of',
  'Minimum of',
  'Maximum of'
}

export type AggregationOperator = keyof typeof AggregationOperatorEnum;

export const aggregationOperators: AggregationOperator[] = [
  'Count of rows',
  'Sum of',
  'Average of',
  'Number of distinct values of',
  'Cumulative sum of',
  'Cumulative count of rows',
  'Standard deviation of',
  'Minimum of',
  'Maximum of'
];
