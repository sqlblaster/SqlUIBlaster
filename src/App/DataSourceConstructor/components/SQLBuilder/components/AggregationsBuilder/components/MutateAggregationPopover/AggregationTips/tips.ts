import { AggregationOperator } from '../../../aggregations';

export const aggregationOperatorToTipsMap: {
  [K in AggregationOperator]: string
} = {
  'Count of rows': 'Total number of rows in the answer',
  'Sum of': 'Sum of the values of a column',
  'Average of': 'Average of the values of a column',
  'Number of distinct values of':
    'Number of unique values of a column among all the rows in the answer',
  'Cumulative sum of':
    'Additive sum of all the values of a column. e.x. total revenue over time',
  'Cumulative count of rows':
    'Additive count of the number of rows. e.x. total number of sales over time',
  'Standard deviation of':
    'Number which expresses how much the value of a column vary among all rows in the answer',
  'Minimum of': 'Minimum value of a column',
  'Maximum of': 'Maximum value of a column'
};

export const getTips = (operator: AggregationOperator): string =>
  aggregationOperatorToTipsMap[operator];
