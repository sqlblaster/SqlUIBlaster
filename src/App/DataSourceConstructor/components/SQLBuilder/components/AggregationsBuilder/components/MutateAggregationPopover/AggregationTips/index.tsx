import {
  createStyles,
  Tooltip,
  withStyles,
  WithStyles
} from '@material-ui/core';
import HelpIcon from '@material-ui/icons/HelpOutline';
import * as React from 'react';
import { AggregationOperator } from '../../../aggregations';
import { getTips } from './tips';

export const styles = () =>
  createStyles({
    maxWidthLimit: {
      maxWidth: 100
    }
  });

export interface AggregationTipsProps extends WithStyles<typeof styles> {
  operator: AggregationOperator;
}

export const AggregationTips = withStyles(styles)(
  ({ operator, classes: { maxWidthLimit } }: AggregationTipsProps) => {
    return (
      <>
        <Tooltip
          title={getTips(operator)}
          placement='bottom'
          classes={{ tooltip: maxWidthLimit }}
        >
          <HelpIcon style={{ color: '#C3C7CA', paddingLeft: '5px' }} />
        </Tooltip>
      </>
    );
  }
);
