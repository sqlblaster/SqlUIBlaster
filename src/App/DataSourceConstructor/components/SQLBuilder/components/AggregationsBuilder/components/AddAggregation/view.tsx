import * as React from 'react';
import { AddEntity } from '../../../common/AddEntity';
import { MutateAggregationPopover } from '../MutateAggregationPopover';
import { AddAggregationViewProps } from './props';

export const addAggregationText = 'Add an aggregation';

// FIXME: make padding clickable too
export class AddAggregation extends React.Component<AddAggregationViewProps, any> {
  private readonly popoverLauncher: React.RefObject<HTMLDivElement>;

  constructor(props: AddAggregationViewProps) {
    super(props);

    this.popoverLauncher = React.createRef<HTMLDivElement>();
  }

  public render() {
    const {
      aggregations,
      addAggregation,
      isOpen,
      handleClose,
      handleOpen
    } = this.props;

    return (
      <>
        <AddEntity
          showText={!(aggregations && aggregations.length)}
          onClick={handleOpen}
          reference={this.popoverLauncher}
        >
          {addAggregationText}
        </AddEntity>
        { isOpen && (// TODO: resolve this problem another way
          <MutateAggregationPopover
            action={addAggregation}
            isOpen={isOpen}
            anchorEl={this.popoverLauncher.current}
            onClose={handleClose}
          />
        ) }
      </>
    );
  }
}
