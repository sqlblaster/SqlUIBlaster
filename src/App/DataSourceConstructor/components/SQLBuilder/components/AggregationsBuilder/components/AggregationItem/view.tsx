import { withStyles } from '@material-ui/core';
import * as React from 'react';
import { ColumnBranch } from 'src/components/query-builder/App/DataSourceConstructor/components/SQLBuilder/components/common/ColumnSelector/ColumnItem/SelectedColumn.models';
import { RemoveButton } from 'src/components/query-builder/App/DataSourceConstructor/components/SQLBuilder/components/common/Icons/RemoveButton';
import { SequentialColumnView } from 'src/components/query-builder/App/DataSourceConstructor/components/SQLBuilder/components/common/SequentialColumnView';
import { MutateAggregationPopover } from '../MutateAggregationPopover';
import { AggregationItemViewProps } from './props';
import { aggregationItemStyles } from './styles';

export const AggregationItem = withStyles(aggregationItemStyles)(
  class extends React.Component<AggregationItemViewProps> {
    private readonly popoverLauncher: React.RefObject<HTMLDivElement>;

    constructor(props: AggregationItemViewProps) {
      super(props);
      this.popoverLauncher = React.createRef<HTMLDivElement>();
    }

    public render() {
      const {
        aggregation,
        updateAggregation,
        classes,
        isOpen,
        handleOpen,
        handleClose,
        handleAggregationRemoval
      } = this.props;

      const { id, operator, column } = aggregation;

      return (
        <>
          <div className={classes.root}>
            <div
              ref={this.popoverLauncher}
              onClick={handleOpen}
              className={classes['column-view']}
            >
              <SequentialColumnView
                prefix={operator}
                {...(column ? column : new ColumnBranch())}
              />
            </div>
            <RemoveButton onClick={handleAggregationRemoval(id)} />
          </div>
          { isOpen && (
            <MutateAggregationPopover
              isOpen={isOpen}
              anchorEl={this.popoverLauncher.current}
              action={updateAggregation}
              onClose={handleClose}
              aggregation={aggregation}
            />
          ) }
        </>
      );
    }
  }
);
