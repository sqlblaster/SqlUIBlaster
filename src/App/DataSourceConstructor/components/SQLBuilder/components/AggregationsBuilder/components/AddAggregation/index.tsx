import * as React from 'react';
import { connect } from 'react-redux';
import { State } from 'src/components/query-builder/store/models/State';
import { WithPopoverManagement } from '../../../common/WithPopoverManagement';
import {
  AAMapDispatchToProps,
  AddAggregationDispatchProps,
  AddAggregationProps,
  AddAggregationStateProps
} from './props';
import { AddAggregation as View } from './view';

export const AddAggregation = connect<
  AddAggregationStateProps,
  AddAggregationDispatchProps,
  {},
  State
>(
  ({ aggregations }) => ({
    aggregations
  }),
  AAMapDispatchToProps
)(
  class extends React.Component<AddAggregationProps> {
    public static ViewWithPopoverManagement = WithPopoverManagement(View);

    public render() {
      const { aggregations, addAggregation } = this.props;

      return (
        <AddAggregation.ViewWithPopoverManagement
          aggregations={aggregations}
          addAggregation={addAggregation}
        />
      );
    }
  }
);
