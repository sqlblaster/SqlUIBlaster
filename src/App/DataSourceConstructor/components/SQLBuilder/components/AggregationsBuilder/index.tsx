import { connect } from 'react-redux';
import { State } from 'src/components/query-builder/store/models/State';
import { AggregationsBuilderStateProps } from './props';
import { AggregationsBuilder as View } from './view';

export const AggregationsBuilder = connect<
  AggregationsBuilderStateProps,
  {},
  {},
  State
>(({ aggregations }) => ({
  aggregations
}))(View);
