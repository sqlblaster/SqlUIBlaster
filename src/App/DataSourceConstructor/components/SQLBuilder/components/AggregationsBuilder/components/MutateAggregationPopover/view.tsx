import { List, ListItem, ListItemText, withStyles } from '@material-ui/core';
import * as React from 'react';
import { ColumnSelector } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/common/ColumnSelector';
import { BackButton } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/common/Icons/BackButton';
import { PopoverPaper } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/common/PopoverPaper';
import { aggregationOperators } from '../../aggregations';
import { Aggregation } from '../AggregationItem/model';
import { AggregationTips } from './AggregationTips';
import { MutateAggregationPopoverViewProps } from './props';
import { mutateAggregationPopoverStyles } from './styles';

export const MutateAggregationPopover = withStyles(
  mutateAggregationPopoverStyles
)((props => {
  const {
    column,
    operator,
    aggregation,
    popoverPositionUpdater,
    classes,
    onColumnSelected,
    onOperatorSelected,
    onBackClicked,
    canShowColumnBranch
  } = props;

  return (
    <PopoverPaper {...props}>
      <div className={classes.root}>
        {((aggregation &&
          Aggregation.hasNoColumn(operator || aggregation.operator)) ||
          !operator) && (
          <div className={classes['select-operator']}>
            <List component='nav' disablePadding={true}>
              {aggregationOperators.map(aggregationOperator => (
                <ListItem
                  key={aggregationOperator}
                  button={true}
                  selected={aggregationOperator === operator}
                  onClick={onOperatorSelected(aggregationOperator)}
                >
                  <ListItemText
                    inset={true}
                    primary={aggregationOperator}
                    style={{ padding: 0 }}
                  />
                  <AggregationTips operator={aggregationOperator} />
                </ListItem>
              ))}
            </List>
          </div>
        )}

        {operator &&
          !Aggregation.hasNoColumn(operator) &&
          ((aggregation && operator && column) || (operator && !column)) && (
            <>
              <div className={classes['title-component']}>
                <BackButton onClick={onBackClicked} />
                {`${operator}...`}
              </div>
              <hr className={classes.hr} />
              <ColumnSelector
                popoverPositionUpdater={popoverPositionUpdater}
                canShowColumnBranch={canShowColumnBranch}
                onColumnSelected={onColumnSelected}
              />
            </>
          )}
      </div>
    </PopoverPaper>
  );
}) as React.FC<MutateAggregationPopoverViewProps>);
