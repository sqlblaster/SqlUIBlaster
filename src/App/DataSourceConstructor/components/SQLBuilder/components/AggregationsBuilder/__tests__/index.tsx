import {
  fireEvent,
  getByText as getByTextOn,
  queryByText as queryByTextOn
} from 'react-testing-library';
import { backIconPlaceholder } from 'src/__tests__/utils/mock-placeholders';
import { regex } from 'src/App/DataSourceConstructor/test-utils/utils';
import { getGeneratedQuery, render } from '../../../__tests__/render.utils';
import { aggregationsBuilderLabel } from '../../../labels';
import { aggregationOperators } from '../../AggregationsBuilder/aggregations';
import {
  clickAddAAggregation,
  clickAddMoreAggregation
} from '../../AggregationsBuilder/components/AddAggregation/__tests__/utils/click-add-a-aggregation';
import { selectAColumn } from '../../common/ColumnSelector/__tests__/utils/select-a-column';
import { selectATable } from '../../TableSelector/__tests__/utils/select-table';
import { addAggregationText } from '../components/AddAggregation/view';
import { getTips } from '../components/MutateAggregationPopover/AggregationTips/tips';

describe('Aggregations Builder', () => {
  const { getByText, queryByText, getAllByText, getByTestId } = render();

  const columns = selectATable();

  const aggregationsBuilderPanel = getByTextOn(
    document.body,
    regex(aggregationsBuilderLabel)
  );

  const clickOnAggregation = (columnName: string) => {
    fireEvent.click(getByTextOn(aggregationsBuilderPanel, regex(columnName)));
  };

  const showsAppropriateLabel = (
    aggregation: typeof aggregationOperators[0]
  ) => {
    const label = getByText(regex(aggregation));

    expect(label.textContent).toMatchSnapshot();
  };

  const removeAnAggregation = () => {
    fireEvent.click(getAllByText(/clear/i)[0]);
  };

  const countOfRowsAggregation = aggregationOperators[0];
  const sumOfAggregation = aggregationOperators[1];
  const numberOfDistinctValuesAggregation = aggregationOperators[3];
  const cumulitiveCountOfRowsAggregation = aggregationOperators[5];
  const aNumberColumn = columns[3];

  const checkNumberTypeAggregation = (
    aggregation: typeof aggregationOperators[0]
  ) => {
    clickAddMoreAggregation();
    fireEvent.click(getByText(aggregation));

    const columnSelectorPanel = getByTestId('column-selector');

    // shows only number columns
    columns.forEach(column => {
      if (column.type === 'number' || column.foreignTableName) {
        expect(
          getByTextOn(columnSelectorPanel, column.name)
        ).toBeInTheDocument();
      } else {
        expect(
          queryByTextOn(columnSelectorPanel, column.name)
        ).not.toBeInTheDocument();
      }
    });

    selectAColumn(aNumberColumn, {
      container: getByTestId('column-selector')
    });

    showsAppropriateLabel(aggregation);

    expect(getGeneratedQuery()).toMatchSnapshot();
    removeAnAggregation();
  };

  it('should show the add aggregation placeholder', () => {
    expect(getByText(addAggregationText)).toBeInTheDocument();
  });

  it('should show the tips messages', () => {
    clickAddAAggregation();

    aggregationOperators.forEach(aggregation => {
      const tips = getTips(aggregation);

      expect(getByText(regex(tips))).toBeInTheDocument();
    });

    fireEvent.click(document.body);
  });

  it(`should add the 'Count of rows' aggregation'`, () => {
    clickAddAAggregation();

    fireEvent.click(getByText(countOfRowsAggregation));

    showsAppropriateLabel(countOfRowsAggregation);

    expect(getGeneratedQuery()).toMatchSnapshot();
  });

  it('should hide the add aggregation placeholder', () => {
    expect(queryByText(addAggregationText)).not.toBeInTheDocument();
  });

  it(`should update aggregation into 'Sum of'`, () => {
    clickOnAggregation(countOfRowsAggregation);
    fireEvent.click(getByText(sumOfAggregation));
    selectAColumn(aNumberColumn);

    showsAppropriateLabel(sumOfAggregation);

    expect(getGeneratedQuery()).toMatchSnapshot();
  });

  it(`should update aggregation into 'Cumulitive count of rows' using back button`, () => {
    clickOnAggregation(sumOfAggregation);
    fireEvent.click(getByText(backIconPlaceholder));
    fireEvent.click(getByText(cumulitiveCountOfRowsAggregation));

    showsAppropriateLabel(cumulitiveCountOfRowsAggregation);

    expect(getGeneratedQuery()).toMatchSnapshot();
  });

  it('should generate indexation for duplicate column aggregations', () => {
    clickAddMoreAggregation();
    fireEvent.click(getAllByText(cumulitiveCountOfRowsAggregation)[1]);

    expect(getGeneratedQuery()).toMatchSnapshot();
  });

  it('should remove aggregation', () => {
    removeAnAggregation();
    removeAnAggregation();

    expect(getGeneratedQuery()).toMatchSnapshot();
  });

  it('should show the add aggregation placeholder again', () => {
    expect(getByText(addAggregationText)).toBeInTheDocument();
  });

  it(`should add 'Number of Distinct values of a date' and apply certain formatting`, () => {
    clickAddMoreAggregation();
    fireEvent.click(getByText(numberOfDistinctValuesAggregation));

    // should show all column types for number of distinct values
    columns.forEach(column => {
      expect(getByText(column.name)).toBeInTheDocument();
    });

    const createdAtColumn = columns[2];
    selectAColumn(createdAtColumn);

    showsAppropriateLabel(numberOfDistinctValuesAggregation);

    expect(getGeneratedQuery()).toMatchSnapshot();
    removeAnAggregation();
  });

  it(`should add 'Average of ' aggregation`, () => {
    const averageOfAggregation = aggregationOperators[2];
    checkNumberTypeAggregation(averageOfAggregation);
  });

  it(`should add 'Cumulative sum of ' aggregation`, () => {
    const cumulitiveSumOfAggregation = aggregationOperators[4];
    checkNumberTypeAggregation(cumulitiveSumOfAggregation);
  });

  it(`should add 'Standard deviation of ' aggregation`, () => {
    const deviationOfAggregation = aggregationOperators[6];
    checkNumberTypeAggregation(deviationOfAggregation);
  });

  it(`should add 'Minimum of ' aggregation`, () => {
    const minimumOfAggregation = aggregationOperators[7];
    checkNumberTypeAggregation(minimumOfAggregation);
  });

  it(`should add 'Maximum of ' aggregation`, () => {
    const maximumOfAggregation = aggregationOperators[8];
    checkNumberTypeAggregation(maximumOfAggregation);
  });

  it(`should generate certain aggregation alias for foreign column`, () => {
    clickAddMoreAggregation();
    fireEvent.click(getByText(sumOfAggregation));

    const productForeignColumn = columns[5];
    const { foreignColumns } = selectAColumn(productForeignColumn, {
      expand: true
    });

    const availableAmountColumn = foreignColumns[2];
    selectAColumn(availableAmountColumn);

    expect(getGeneratedQuery()).toMatchSnapshot();
  });
});
