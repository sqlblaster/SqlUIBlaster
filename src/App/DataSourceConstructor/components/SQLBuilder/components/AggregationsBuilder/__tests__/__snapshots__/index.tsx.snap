// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Aggregations Builder should add 'Average of ' aggregation 1`] = `"Average ofQuantity"`;

exports[`Aggregations Builder should add 'Average of ' aggregation 2`] = `
"SELECT
  avg(\\"orders1\\".\\"Quantity\\") AS \\"Average of Quantity\\"
FROM
  \\"orders\\" \\"orders1\\"
LIMIT 2000
"
`;

exports[`Aggregations Builder should add 'Cumulative sum of ' aggregation 1`] = `"Cumulative sum ofQuantity"`;

exports[`Aggregations Builder should add 'Cumulative sum of ' aggregation 2`] = `
"SELECT
  sum(\\"orders1\\".\\"Quantity\\") AS \\"Cumulative sum of Quantity\\"
FROM
  \\"orders\\" \\"orders1\\"
LIMIT 2000
"
`;

exports[`Aggregations Builder should add 'Maximum of ' aggregation 1`] = `"Maximum ofQuantity"`;

exports[`Aggregations Builder should add 'Maximum of ' aggregation 2`] = `
"SELECT
  max(\\"orders1\\".\\"Quantity\\") AS \\"Maximum of Quantity\\"
FROM
  \\"orders\\" \\"orders1\\"
LIMIT 2000
"
`;

exports[`Aggregations Builder should add 'Minimum of ' aggregation 1`] = `"Minimum ofQuantity"`;

exports[`Aggregations Builder should add 'Minimum of ' aggregation 2`] = `
"SELECT
  min(\\"orders1\\".\\"Quantity\\") AS \\"Minimum of Quantity\\"
FROM
  \\"orders\\" \\"orders1\\"
LIMIT 2000
"
`;

exports[`Aggregations Builder should add 'Number of Distinct values of a date' and apply certain formatting 1`] = `"Number of distinct values ofCreated At"`;

exports[`Aggregations Builder should add 'Number of Distinct values of a date' and apply certain formatting 2`] = `
"SELECT
  count(DISTINCT date_trunc('day', \\"orders1\\".\\"Created At\\")) AS \\"Distinct count of Created At\\"
FROM
  \\"orders\\" \\"orders1\\"
LIMIT 2000
"
`;

exports[`Aggregations Builder should add 'Standard deviation of ' aggregation 1`] = `"Standard deviation ofQuantity"`;

exports[`Aggregations Builder should add 'Standard deviation of ' aggregation 2`] = `
"SELECT
  stddev(\\"orders1\\".\\"Quantity\\") AS \\"Standard deviation of Quantity\\"
FROM
  \\"orders\\" \\"orders1\\"
LIMIT 2000
"
`;

exports[`Aggregations Builder should add the 'Count of rows' aggregation' 1`] = `"Count of rows"`;

exports[`Aggregations Builder should add the 'Count of rows' aggregation' 2`] = `
"SELECT
  count(*) AS \\"Count of rows\\"
FROM
  \\"orders\\"
LIMIT 2000
"
`;

exports[`Aggregations Builder should generate certain aggregation alias for foreign column 1`] = `
"SELECT
  sum(\\"products1\\".\\"Available amount\\") AS \\"Sum of Available amount (products1)\\"
FROM
  \\"orders\\" \\"orders1\\"
  LEFT JOIN \\"products\\" \\"products1\\" ON \\"orders1\\".\\"ProductID\\" = \\"products1\\".\\"ID\\"
LIMIT 2000
"
`;

exports[`Aggregations Builder should generate indexation for duplicate column aggregations 1`] = `
"SELECT
  count(*) AS \\"Cumulative count of rows\\",
  count(*) AS \\"Cumulative count of rows1\\"
FROM
  \\"orders\\"
LIMIT 2000
"
`;

exports[`Aggregations Builder should remove aggregation 1`] = `
"SELECT
  *
FROM
  \\"orders\\"
LIMIT 2000
"
`;

exports[`Aggregations Builder should update aggregation into 'Cumulitive count of rows' using back button 1`] = `"Cumulative count of rows"`;

exports[`Aggregations Builder should update aggregation into 'Cumulitive count of rows' using back button 2`] = `
"SELECT
  count(*) AS \\"Cumulative count of rows\\"
FROM
  \\"orders\\"
LIMIT 2000
"
`;

exports[`Aggregations Builder should update aggregation into 'Sum of' 1`] = `"Sum ofQuantity"`;

exports[`Aggregations Builder should update aggregation into 'Sum of' 2`] = `
"SELECT
  sum(\\"orders1\\".\\"Quantity\\") AS \\"Sum of Quantity\\"
FROM
  \\"orders\\" \\"orders1\\"
LIMIT 2000
"
`;
